### PR TITLE
[MRG] Jail

### PR DESCRIPTION
**Pull request by aapeliv**
As discussed on Slack, I've implemented a prototype "jail" idea for review/comments. I'm not sure if this is a good way of solving these problems, so I'd like some feedback on what people think. It feels kind of convoluted and especially tacking it on now to the authentication system (that assumed a 2-state model) makes it a bit awkward (there's now 3 states: not logged in, logged in, and logged in but jailed).

This makes it possible for a user to be authenticated while not being able to use the main APIs, but only a subset, the "jailed" ones. This is useful if:

1. The user has not finished signing up (e.g. has not finished adding a location)
2. The user has not accepted TOS/GDPR/etc
3. The user is banned, but we'd like to show them something nicer than a "you're banned" at login

It makes it easy to add requirements, e.g. update a privacy policy and make sure everyone has agreed to them, or add required fields to a user profile.

This is also missing tests.

### CODE REVIEW

File: app/backend/src/couchers/models.py
Comment (kalvdans) [inline app/backend/src/couchers/models.py]: Add more indices to the database instead if you worry about performance. No duplicated state please!

File: app/backend/src/couchers/interceptors.py
Comment (kalvdans) [inline app/backend/src/couchers/interceptors.py]: Maybe don't leak the word Jail to the end user. (some clients may show the error reason straight up)

**Review by kalvdans — CHANGES_REQUESTED** [review]
Please versionify TOS (int instead of bool) to easier allow new versions of TOS to be deployed without updating the database.

File: app/backend/src/couchers/interceptors.py
Comment (aapeliv) [reply app/backend/src/couchers/interceptors.py]: We should pick some appropriate word.

**Comment by aapeliv** [comment]
I've changed the TOS field to an integer, and removed the duplicated state. I haven't added any logic to actually version the TOS field, I'll leave that to later. I wasn't actually thinking yet as far as to put some real TOS, it was just a "sample placeholder" jail thing. As @NightTsarina suggested, I'm trying to keep PRs smaller, and wanted feedback on jail before slapping the user location on top of it.

File: app/backend/src/couchers/servicers/auth.py
Comment (kalvdans) [inline app/backend/src/couchers/servicers/auth.py]: Stale comment, remove.

**Review by kalvdans — CHANGES_REQUESTED** [review]
Please add a comment in the top of the the unjailed API descriptions files (api.proto, conversations.proto, media.proto, requests.proto). Something along these lines:
```
// All these calls can generate Unauthenticated Error for the following reasons: 
// Session token expired, user blocked, TOS changed or New profile data needed.
// Call `JailInfo()` to get the exact reason.
```

**Comment by aapeliv** [comment]
I added tests. This is almost ready, except the `TOS.vue` page should probably say something a bit more elaborate than `I understand Couchers.org is not a dating platform, etc, etc.`.

Also, I hope to merge #319 and update migrations before merging this.

**Comment by aapeliv** [comment]
How about "paddock", "cage", or just "box"?

**Comment by lucaslcode** [comment]
How about "pending"? Like a pending user.

File: app/backend/src/couchers/models.py
Comment (lucaslcode) [inline app/backend/src/couchers/models.py:116]: Should we have '1' stored in config somewhere as the current TOS version?

File: app/backend/src/tests/test_jail.py
Comment (lucaslcode) [inline app/backend/src/tests/test_jail.py]: false?

File: app/vue/src/views/TOS.vue
Comment (lucaslcode) [inline app/vue/src/views/TOS.vue]:
```suggestion
        >{{ acceptedTos ? "Already accepted" : "Accept" }}</v-btn
```

File: app/vue/src/views/TOS.vue
Comment (lucaslcode) [inline app/vue/src/views/TOS.vue]:
```suggestion
      <p>This is a test TOS. I agree to follow the <a href="https://community.couchers.org/faq">Couchers guidelines</a>. I understand this is a preview of couchers.org and my data may be erased.</p>
```

File: app/backend/src/tests/test_fixtures.py
Comment (lucaslcode) [inline app/backend/src/tests/test_fixtures.py]: This could make testing of other jailing reasons difficult. Maybe for now you could change the argument from 'jailed' to 'acceptedTOS'?

**Review by lucaslcode — DISMISSED** [review]
Only minor nitpicks

File: app/backend/src/couchers/models.py
Comment (aapeliv) [reply app/backend/src/couchers/models.py:116]: Yes, there should be some logic for this, but I didn't want to do it yet as it's not very relevant yet.

Actually, maybe all it takes is that when the TOS are updated, we do an `UPDATE users SET accepted_tos = 0` :P

File: app/backend/src/tests/test_jail.py
Comment (aapeliv) [reply app/backend/src/tests/test_jail.py]: Yep, you're right

File: app/vue/src/views/TOS.vue
Comment (aapeliv) [reply app/vue/src/views/TOS.vue]: Thanks

File: app/backend/src/tests/test_fixtures.py
Comment (aapeliv) [reply app/backend/src/tests/test_fixtures.py]: Actually, on a second thought, I don't think this should be a parameter here anyway, the jail test modifies the user as needed anyway. I'll refactor it out.

File: app/backend/src/couchers/models.py
Comment (kalvdans) [reply app/backend/src/couchers/models.py:116]: The frontend needs to pass in which version of the TOS the user has agreed to.

**Comment by aapeliv** [comment]
@lucaslcode: does this require some quick patch on the react frontend so signup still works and it's not bricked?

**Comment by lucaslcode** [comment]
On it #342

File: app/backend/src/couchers/models.py
Comment (aapeliv) [reply app/backend/src/couchers/models.py:116]: @kalvdans: could we leave the rest of versioned TOS to a later feature?

Comment (kalvdans) [reply app/backend/src/couchers/models.py:116]: Sure, at the expense of more database migrations/API  incompatibilities.

**Review by kalvdans — DISMISSED** [review]
Thanks for documenting the calling contexts in the API descriptions! LGTM!

**Comment by lucaslcode** [comment]
@aapeliv I have to resolve these merge frontend conflicts but there's scary alembic errors, what's the best thing to do? Merge develop into this branch and let you fix alembic?

**Comment by lucaslcode** [comment]
I made the commit, feel free to revert it

**Comment by aapeliv** [comment]
> @aapeliv I have to resolve these merge frontend conflicts but there's scary alembic errors, what's the best thing to do? Merge develop into this branch and let you fix alembic?

Thanks, I'll fix it and merge

**Pull request merged**
